models/nft/ethereum/nft_ethereum_wallet_pnl.sql

```diff
@@ -0,0 +1,149 @@
+        -1 * SUM(amount_original) as eth_amount,
+        COUNT(*) as trades, 
+        NOW() - Interval '5 Minutes' as last_updated
+    FROM 
+    {{ ref('nft_trades') }}
+    WHERE currency_symbol IN ('ETH', 'WETH')
+    AND blockchain = 'ethereum'
+    AND buyer != seller 
+    AND trade_type = 'Single Item Trade'
+    AND amount_original IS NOT NULL 
+    AND block_time < NOW() - Interval '5 Minutes'
+    GROUP BY 1, 2, 3, 6
+    {% endif %}
+    -- incremental run 
+    {% if is_incremental() %}
+        seller as wallet, 
+        nft_contract_address, 
+        'Sells' as trade_type,
+        SUM(amount_original) as eth_amount, 
+        COUNT(*) as trades, 
+        NOW() - Interval '5 Minutes' as last_updated
+    FROM 
+    {{ ref('nft_trades') }}
+    WHERE block_time >= (SELECT MAX(last_updated) - Interval '5 Minutes' FROM {{this}})
+    AND currency_symbol IN ('ETH', 'WETH')
+    AND blockchain = 'ethereum'
+    AND buyer != seller 
+    AND trade_type = 'Single Item Trade'
+    AND amount_original IS NOT NULL 
+    AND block_time < NOW() - Interval '5 Minutes'
+    GROUP BY 1, 2, 3, 6
+
+    UNION ALL 
+
+    SELECT 
+        buyer as wallet, 
+        nft_contract_address, 
+        'Buys' as trade_type,
+        -1 * SUM(amount_original) as eth_amount,
+        COUNT(*) as trades, 
+        NOW() - Interval '5 Minutes' as last_updated
+    FROM 
+    {{ ref('nft_trades') }}
+    WHERE block_time >= (SELECT MAX(last_updated) FROM {{this}})
+    AND currency_symbol IN ('ETH', 'WETH')
+    AND blockchain = 'ethereum'
+    AND buyer != seller 
+    AND trade_type = 'Single Item Trade'
+    AND amount_original IS NOT NULL 
+    AND block_time < NOW() - Interval '5 Minutes'
+    GROUP BY 1, 2, 3, 6
+    {% endif %}
+), 
+
+combined as (
+    SELECT 
+     {% if not is_incremental() %}
+        wallet, 
+        nft_contract_address, 
+        last_updated,
+        COALESCE(SUM(CASE 
+            WHEN trade_type = 'Buys'
+            THEN ABS(eth_amount) 
+            ELSE 0 
+        END), 0) as eth_spent, 
+        COALESCE(SUM(CASE 
+            WHEN trade_type = 'Sells'
+            THEN eth_amount
+            ELSE 0 
+        END), 0) as eth_received, 
+        SUM(eth_amount) as pnl, 
+        SUM(trades) as trades 
+    FROM 
+    trades 
+    GROUP BY 1, 2, 3
+    {% endif %}
+    {% if is_incremental() %}
+    SELECT 
+        prev.wallet, 
+        prev.nft_contract_address,
+        t.last_updated,
+        COALESCE(SUM(CASE 
+            WHEN t.trade_type = 'Buys'
+            THEN ABS(t.eth_amount) 
+            ELSE 0 
+        END), 0) + prev.eth_spent as eth_spent,
+        COALESCE(SUM(CASE 
+            WHEN t.trade_type = 'Sells'
+            THEN t.eth_amount
+            ELSE 0 
+        END), 0) + prev.eth_received as eth_received,
+        COALESCE(SUM(t.eth_amount), 0) +  prev.pnl as pnl, 
+        COALESCE(SUM(t.trades), 0) + prev.trades as trades 
+    FROM 
+    {{this}} prev 
+    LEFT JOIN 
```

Comment:
I see this will mean we will be working with the entire history of the dataset every time we do an incremental update, do you need this for anything other than calculating the rank at the end?

If we can do this as an inner join instead I think it would be best, so we're not running a merge with all previous records.

I may be thin on the performance optimisation details here so if you guys know better please correct me.

Reply:
hmm, I'm happy to remove the index. But the reason we're using a left join is to add up the partial pnl. At the time I wrote this incremental I had not written for the top sales & top minters so I think it's better to use a UNION ALL here?

So instead of left join, the incremental query runs a UNION ALL of SELECT * FROM {{this}} and then unions that with all the new entries after the max(last_updated)?

Reply:
Also, it doesn't actually look through the entire history of the dataset, well that's what I think I have written but it's my first time playing with incremental like this. 

But the only time it goes through the dataset entire history is the first run? and in the incremental run, it just takes the previous results of the table ({{this}}) and then adds the pnl from the previous run to the results after (max(last_updated))

Reply:
bump ser @augustog 👀

Reply:
Sorry for the slow response! 

1. my suggestion to use inner join is indeed wrong, sorry for the noise there! If we used inner join here we would never add new wallets to the dataset
2. my point was, however, that if we select from {{this}} without any filters, we'll pull all records the spell ever created, change a few (the ones with new trades => different pnl) and then merge into the existing dataset, having to drop most of the records (as they will be the same). This is what I meant with "scanning the whole history of the dataset" (sorry if it was not very clear).

My advise would be to turn this into a RIGHT JOIN. This would mean we pull all new trades since last run, and all the PnLs that match those trades by wallet and collection) I'm not sure whether this would improve performance of the join itself too much, but at least we won't be generating a copy of the dataset for every update.

Does this make sense?

Reply:
An alternative would be to use a simple query that sum's the pnl of traders that's grouped and partitioned by block_date. 

And then do the SUM(pnl) when the spell is queried on dune.com. I'm not sure which is better though but just something I thought of.

Reply:
i think the issue here is if you read from {{this}} object and left join new incremental data, it'll only ever contain the wallets from the initial run. it won't add in any new wallets, as the left join keeps only from {{this}} model and corresponding matches from incremental.

my suggestion would be either keep it simple and make this just a table with no incremental logic -- CI test shows it runs pretty quickly and will just full refresh each time -- or potentially switch from left join to full outer join, to get wallets from both sides, then the logic to add across both tables should work.

this is a bit difficult to test on live prod data sources, it'd be good if we could replicate data to play out these scenarios.

Reply:
hmm, ohh you're right ser. 

What do you think about the **UNION ALL** approach? Would that be simpler / faster?

The issue is that there's no reason to have this on spellbook if it's going to have to go through the entire nft.trades table during every run. That way it can just be queried from dune.com? Whilst the incremental run would only sum up the previous data + new data.

Reply:
bump ser @jeff-dude